keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+
+## Design Details
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
+
+    // This defines which nodes the config is applicable to. A nil selector can
+    // be applied to any node.
+    // +optional
+    NodeSelector *v1.LabelSelector
+
+    // Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    // Users would have to ensure that the kubelet doesn't try to schedule
+    // more pods than are supported by the node's netmask (i.e. the kubelet's
+    // --max-pods flag)
+    // +required
+    PerNodeMaskSize int
+
+    // Is this a good idea?
+    //
+    // If Disabled is set to false, no new PodCIDRs will be allocated from this range
+    // +optional
+    Disabled bool
+}
+
+type ClusterCIDRConfigStatus {
+  Conditions []metav1.Conditions
+}
+
+type ClusterCIDRConfigStatusType string
+type ClusterCIDRConfigStatusReason string
+var (
+  // If the "active" condition is true, this ClusterCIDRConfig can be used by
+  // the controller to allocate PodCIDRs for matching nodes.
+  ClusterCIDRConfigActive ClusterCIDRConfigStatusType = "active"
+
+  // If the "terminating" condition is true, this ClusterCIDRConfig was deleted
+  // by a user and is being garbage collected. When all Nodes using PodCIDRs
+  // from this range are deleted, the ClusterCIDRConfig will also be deleted.
+  ClusterCIDRConfigTerminating ClusterCIDRConfigStatusType = "terminating"
+
+  // If set as the reason on a false ClusterCIDRConfigActive condition, the
+  // ClusterCIDRConfig no longer has any free IP blocks.
+  ClusterCIDRConfigExhausted ClusterCIDRConfigStatusReason = "cidr-exhausted"
+
+  // If set as the reason on a true ClusterCIDRConfigTerminating condition,
+  // the ClusterCIDRConfig was used to allocate a Node's PodCIDR.
+  ClusterCIDRConfigInUse ClusterCIDRConfigStatusReason "cidr-in-use"
+)
+
+```
+
+#### Expected Behavior
+
+-   Each node will be assigned up to one range from each `FamilyType`. In case
+    of multiple matching ranges, attempt to break ties with the following rules:
+    1.  Pick the `ClusterCIDRRange` whose `NodeSelector` matches the most
+        labels on the `Node`. For example,
+        `{'node.kubernetes.io/instance-type': 'medium', 'rack': 'rack1'}`
+        before `{'node.kubernetes.io/instance-type': 'medium'}`.
+    1.  Break ties arbitrarily.
+
+-   An empty `NodeSelector` functions as a default that applies to all nodes.
+    This should be the fall-back and not take precedence if any other range
+    matches. If there are multiple default ranges, ties are broken using the
+    scheme outlined above.
+
+-   `CIDR` and `PerNodeMaskSize` are immutable after creation.
+
+-   The controller will add a finalizer to the ClusterCIDRConfig object when it
```

Comment:
I'm not convinced the conditions are needed

Reply:
Dropped the conditions.